MINOR: Add logging to Connect SMTs

cyrusv opened the pull request:
Includes Update to ConnectRecord string representation to give
visibility into schemas, useful in SMT tracing

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/TransformationChain.java:
```diff
@@ -68,4 +68,14 @@ public int hashCode() {
 
+    public String toString() {
+        StringBuilder str = new StringBuilder();
+        str.append(getClass().getName());
```
How about short name for `TransformChain`?

cyrusv replied:
I'm not sure what you mean?

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/TransformationChain.java:
```diff
@@ -68,4 +68,14 @@ public int hashCode() {
+        str.append(": ,");
+        for (final Transformation<R> transformation : transformations) {
+            str.append(transformation.getClass().getName());
+            str.append(", ");
```
This will end with a comma after the last transformation.

rhauch commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/Cast.java:
```diff
@@ -102,8 +105,10 @@ public void configure(Map<String, ?> props) {
     @Override
     public R apply(R record) {
         if (operatingSchema(record) == null) {
+            log.trace("Applying Cast SMT with null schema. Record: {}", record);
```
Would it be better to log these traces inside `TransformChain`, where we could see the progression of the records through the chain?

cyrusv replied:
My original idea was that this gives some more useful information about which branch we are taking. Given your comment and a second thought, I think we a log in the Chain since we can't count on every SMT having the right logs, and it will provide consistency.

cyrusv replied:
I will defer to your judgement here since I haven't debugged one of these in production before -- new strategy will be to put log in TransformChain.apply() for each transformation, and then put some smaller trace logs in transformations that distinguish between schema'd and schemaless. If the schema/schemaless distinction isn't very useful, I can strike that

rhauch commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/InsertField.java:
```diff
@@ -128,8 +131,10 @@ public void configure(Map<String, ?> props) {
+            log.trace("Applying InsertField SMT with null schema. Record: {}", record);
             return applySchemaless(record);
         } else {
+            log.trace("Applying InsertField SMT with schema. Record: {}", record);
```
Is it clear that the record is the _before_ record? Maybe "Applying InsertField SMT with schema to record {}" instead?

rhauch reviewed (commented):
Thanks, @cyrusv. Left a few comments / suggestions.

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java:
```diff
@@ -493,6 +493,7 @@ private WorkerTask buildWorkerTask(ClusterConfigState configState,
         if (task instanceof SourceTask) {
             retryWithToleranceOperator.reporters(sourceTaskReporters(id, connConfig, errorHandlingMetrics));
             TransformationChain<SourceRecord> transformationChain = new TransformationChain<>(connConfig.<SourceRecord>transformations(), retryWithToleranceOperator);
+            log.debug("Initializing: {}", transformationChain);
```
I'm actually wondering if this should be INFO level, just so that it's clear what SMTs are actually being used for this connector, in case the user messed up the configuration. Thoughts? Is this logged anywhere else?

cyrusv replied:
agreed

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java:
```diff
@@ -505,6 +506,7 @@ private WorkerTask buildWorkerTask(ClusterConfigState configState,
                     time, retryWithToleranceOperator);
         } else if (task instanceof SinkTask) {
             TransformationChain<SinkRecord> transformationChain = new TransformationChain<>(connConfig.<SinkRecord>transformations(), retryWithToleranceOperator);
+            log.debug("Initializing: {}", transformationChain);
```
Same question as above.

rhauch commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/Cast.java:
```diff
@@ -154,10 +158,14 @@ private R applyWithSchema(R record) {
+            log.trace("Cast SMT: Original value: {}", origFieldValue);
             final Schema.Type targetType = casts.get(field.name());
             final Object newFieldValue = targetType != null ? castValueToType(field.schema(), origFieldValue, targetType) : origFieldValue;
+            log.trace("Cast SMT: New value: {}", newFieldValue);
```
This is a useful log message. But since in a busy Connect worker it's unlikely these log two messages will be adjacent, how about instead using a single log message:

    log.trace("Cast field '{}' from '{}' to '{}'", field.name(), origFieldValue, newFieldValue);

rhauch commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/Cast.java:
```diff
@@ -234,6 +243,7 @@ private static Object castValueToType(Schema schema, Object value, Schema.Type t
             // Ensure the type we are trying to cast from is supported
             validCastType(inferredType, FieldType.INPUT);
 
+            log.trace("Casting to {}", targetType);
```
I'm not sure this is worth it, if we're logging messages for each field being converted, plus if the TransformationChain has log messages with the records.

rhauch commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/Cast.java:
```diff
@@ -188,6 +196,7 @@ private Schema getOrBuildSchema(Schema valueSchema) {
         }
 
         if (valueSchema.isOptional())
+            log.trace("Cast SMT: ValueSchema is optional");
```
This probably doesn't add value without more information about the name of the schema.

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/TransformationChain.java:
```diff
@@ -40,6 +44,8 @@ public R apply(R record) {
         for (final Transformation<R> transformation : transformations) {
             final R current = record;
 
+            log.trace("Applying SMT transformation: {}, record: {}",
```
These log messages really make most of the individual log messages within the SMTs less valuable, since the record will include the key schema, key, value schema, and value.

I've highlighted some of the log messages in individual SMTs that I think are worthwhile.

Also: 
```suggestion
            log.trace("Applying SMT transformation {} to {}",
```

rhauch replied:
Also, I just looked at `ConnectRecord.toString()`, and it does _not_ print the schemas. I wonder if it's worth changing that to include whether the key schema and value schema are null; e.g.,
```
    @Override
    public String toString() {
        return "ConnectRecord{" +
                "topic='" + topic + '\'' +
                ", kafkaPartition=" + kafkaPartition +
                ", keySchema=" + (keySchema != null ? "..." : "null") +
                ", key=" + key +
                ", valueSchema=" + (valueSchema != null ? "..." : "null") +
                ", value=" + value +
                ", timestamp=" + timestamp +
                ", headers=" + headers +
                '}';
    }
```

rhauch commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/Cast.java:
```diff
@@ -137,6 +140,7 @@ private R applySchemaless(R record) {
             String field = fieldSpec.getKey();
             updatedValue.put(field, castValueToType(null, value.get(field), fieldSpec.getValue()));
         }
+        log.trace("Schemaless SMT application with updatedValue: {}", updatedValue);
```
I'm not sure these add value since TransformationChain has log messages with the records, which should be sufficient to know whether the schemas are null.

rhauch commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/Cast.java:
```diff
@@ -154,10 +158,14 @@ private R applyWithSchema(R record) {
             updatedValue.put(updatedSchema.field(field.name()), newFieldValue);
         }
+        log.trace("Schema SMT application with updatedValue: {}, updatedSchema: {}",
+            updatedValue, updatedSchema);
```
I'm not sure these add value since TransformationChain has log messages with the records, which should be sufficient to know whether the schemas are null.

rhauch commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/ExtractField.java:
```diff
@@ -21,13 +21,16 @@
 import static org.apache.kafka.connect.transforms.util.Requirements.requireStructOrNull;
 
 public abstract class ExtractField<R extends ConnectRecord<R>> implements Transformation<R> {
+    private static final Logger log = LoggerFactory.getLogger(ExtractField.class);
```
I'm not sure any of these log messages add value since TransformationChain has log messages with the records, which should be sufficient to know whether the schemas are null.

rhauch commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/Flatten.java:
```diff
@@ -38,6 +40,7 @@
 import static org.apache.kafka.connect.transforms.util.Requirements.requireStruct;
 
 public abstract class Flatten<R extends ConnectRecord<R>> implements Transformation<R> {
+    private static final Logger log = LoggerFactory.getLogger(Flatten.class);
```
I'm not sure any of these log messages add value since TransformationChain has log messages with the records, which should be sufficient to know whether the schemas are null.

rhauch commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/HoistField.java:
```diff
@@ -25,11 +25,14 @@
 import java.util.Map;
 
 public abstract class HoistField<R extends ConnectRecord<R>> implements Transformation<R> {
+    private static final Logger log = LoggerFactory.getLogger(HoistField.class);
```
I'm not sure any of these log messages add value since TransformationChain has log messages with the records, which should be sufficient to know whether the schemas are null.

rhauch commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/InsertField.java:
```diff
@@ -39,6 +41,7 @@
 import static org.apache.kafka.connect.transforms.util.Requirements.requireStruct;
 
 public abstract class InsertField<R extends ConnectRecord<R>> implements Transformation<R> {
+    private static final Logger log = LoggerFactory.getLogger(InsertField.class);
```
I'm not sure any of these log messages add value since TransformationChain has log messages with the records, which should be sufficient to know whether the schemas are null.

rhauch commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/MaskField.java:
```diff
@@ -39,6 +41,7 @@
 import static org.apache.kafka.connect.transforms.util.Requirements.requireStruct;
 
 public abstract class MaskField<R extends ConnectRecord<R>> implements Transformation<R> {
+    private static final Logger log = LoggerFactory.getLogger(MaskField.class);
```
I'm not sure any of these log messages add value since TransformationChain has log messages with the records, which should be sufficient to know whether the schemas are null.

rhauch commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/RegexRouter.java:
```diff
@@ -20,12 +20,15 @@
 import java.util.regex.Pattern;
 
 public class RegexRouter<R extends ConnectRecord<R>> implements Transformation<R> {
+    private static final Logger log = LoggerFactory.getLogger(RegexRouter.class);
```
I'm not sure any of these log messages add value since TransformationChain has log messages with the records, which should be sufficient to know whether the schemas are null.

rhauch commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/ReplaceField.java:
```diff
@@ -38,6 +40,7 @@
 import static org.apache.kafka.connect.transforms.util.Requirements.requireStruct;
 
 public abstract class ReplaceField<R extends ConnectRecord<R>> implements Transformation<R> {
+    private static final Logger log = LoggerFactory.getLogger(ReplaceField.class);
```
I'm not sure any of these log messages add value since TransformationChain has log messages with the records, which should be sufficient to know whether the schemas are null.

rhauch commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/SetSchemaMetadata.java:
```diff
@@ -76,6 +79,8 @@ public R apply(R record) {
                 isMap || isArray ? schema.valueSchema() : null
         );
+        log.trace("Applying SetSchemaMetadata SMT. Original schema: {}, updated schema: {}",
+            schema, updatedSchema);
```
This may be useful, since the log messages in TransformationChain do not print the schemas in `ConnectRecord`

rhauch commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/TimestampConverter.java:
```diff
@@ -50,6 +52,7 @@
 import static org.apache.kafka.connect.transforms.util.Requirements.requireStruct;
 
 public abstract class TimestampConverter<R extends ConnectRecord<R>> implements Transformation<R> {
+    private static final Logger log = LoggerFactory.getLogger(TimestampConverter.class);
```
I'm not sure any of these log messages add value since TransformationChain has log messages with the records, which should be sufficient to know whether the schemas are null.

rhauch commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/TimestampRouter.java:
```diff
@@ -29,6 +31,7 @@
 import java.util.regex.Pattern;
 
 public class TimestampRouter<R extends ConnectRecord<R>> implements Transformation<R> {
+    private static final Logger log = LoggerFactory.getLogger(TimestampRouter.class);
```
I'm not sure any of these log messages add value since TransformationChain has log messages with the records, which should be sufficient to know whether the schemas are null.

rhauch commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/ValueToKey.java:
```diff
@@ -35,6 +37,7 @@
 import static org.apache.kafka.connect.transforms.util.Requirements.requireStruct;
 
 public class ValueToKey<R extends ConnectRecord<R>> implements Transformation<R> {
+    private static final Logger log = LoggerFactory.getLogger(ValueToKey.class);
```
This is unused.

rhauch reviewed (commented):
Thanks, @cyrusv! I think we can simplify this dramatically by focusing on good trace log messages in `TransformationChain`.

cyrusv committed:
MINOR: Add logging to Connect SMTs

More trace-level logging will improve insight into SMIT chain

Includes Update to ConnectRecord string representation to give
visibility into schemas, useful in SMT tracing

rhauch commented on connect/api/src/main/java/org/apache/kafka/connect/connector/ConnectRecord.java:
```diff
@@ -141,6 +141,7 @@ public String toString() {
                 ", kafkaPartition=" + kafkaPartition +
                 ", key=" + key +
                 ", value=" + value +
+                ", valueSchema=" + (valueSchema != null ? "..." : "null") +
```
Why not the key schema as well?

rhauch reviewed (commented):
One more request. Otherwise, looks great.

rhauch reviewed (approved):
LGTM. Thanks!

kkonstantine commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/TransformationChain.java:
```diff
@@ -40,6 +44,8 @@ public R apply(R record) {
         for (final Transformation<R> transformation : transformations) {
             final R current = record;
 
+            log.trace("Applying SMT transformation {} to {}",
```
There's some redundancy here. SMT=Single Message Transform + transformation
Given that you log the name of the transformation I'd use `"Applying transformation {}  to {}"`

cyrusv replied:
I considered this "ATM Machine" effect when I wrote this, but decided that putting every name for something helps with readability more than grammar does. It also will help with grepping/searching the code base. If you feel strongly, I think a strong compromise would be "Applying Single Message Transformation (SMT)..."

kkonstantine replied:
I don't see how wrong usage elsewhere inspires wrong usage here. 
Despite that, I still prefer `Applying transformation` since this log message is not a paragraph in a text where we need to be precise and need to list the acronym explicitly along with what it means. This is a log message that needs to be succinct and can be read in context. A transformation here implies a connect SMT.

kkonstantine commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/TransformationChain.java:
```diff
@@ -40,6 +44,8 @@ public R apply(R record) {
             final R current = record;
 
+            log.trace("Applying SMT transformation {} to {}",
+                transformation.getClass().getName(), record);
```
Are you sure you want the FQCN here? simple name is another option which might be nicer to print. But in any case, want to make sure we are deliberate here.

kkonstantine replied:
(especially given that below you use the simple name)

kkonstantine commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/TransformationChain.java:
```diff
@@ -68,4 +74,11 @@ public int hashCode() {
     }
 
+    public String toString() {
+        StringJoiner str = new StringJoiner(", ", "<" + getClass().getSimpleName() + ": ", ">");
```
I've found that avoiding using anything related to the type of a variable in its name is usually for the best. It takes a few more seconds to come up with an appropriate name, but it definitely pays off when reading, and even in generic uses as this one here it's still better IMO. E.g. I'd call this at least `chain` or something.

cyrusv replied:
I don't understand your comment -- can you explain what you mean by "avoiding using anything related to the type of a variable"?

cyrusv replied:
Oh, you mean the variable name "str"?

kkonstantine commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/TransformationChain.java:
```diff
@@ -68,4 +74,11 @@ public int hashCode() {
 
+    public String toString() {
+        StringJoiner str = new StringJoiner(", ", "<" + getClass().getSimpleName() + ": ", ">");
+        for (final Transformation<R> transformation : transformations) {
```
I'm against the sprinkling of `final` keyword everywhere. But more than that, I'm in favor of consistency w.r.t. the surrounding code. A few lines above, you may observe that `close` has a similar loop without `final`. Every project has its idiosyncrasies and Connect is not dogmatic w.r.t to `final` for local variables.

kkonstantine replied:
Also, am I correct to read that every transformation name will be prefixed by this classes simple classname as well. Is this the visual result we want the most? Can we get an example as a non-javadoc comment here, if it's useful?

cyrusv replied:
I agree with consistency re: `final`. 

I'm not sure where your confusion re:simpleName is coming from. I hesitate to put an example of the output in the comments since it could lead to future inconsistency, but the intention is to put simple name of each transform with the record it will be applied to. I will paste a sample log output below

cyrusv replied:
[2018-11-13 16:06:08,083] TRACE Registered Applying Single Message Transformation (SMT) HoistField to SinkRecord{kafkaOffset=45, timestampType=NoTimestampType} ConnectRecord{topic='test', kafkaPartition=12, key=null, keySchema=null, value={a=10}, valueSchema=null, timestamp=null, headers=ConnectHeaders(headers=)} (org.apache.kafka.connect.runtime.TransformationChain:47)

kkonstantine replied:
I don't feel too strong about the comment above the code, even though I don't think there's an issue of that becoming stale. It's what this method does and should be updated when (infrequently) this method changes, if ever. 

But in your example above, I don't see neither the prefix nor the suffix that are given to the `StringJoiner`

kkonstantine replied:
@cyrusv 
Your example should print something like: 
`<SimpleClassName: blah blah blah>`
But I don't see `<` or `>` in your example. So either the code example is wrong, or you didn't mean to use those "quotes" in the joiner. BTW probably not adding this decoration would be better.

cyrusv replied:
@kkonstantine  I see your point -- I updated the code to match the style used in `ConnectRecord.toString()`, so now the output will look like "TransformationChain{FaultyPassthrough, FaultyPassthrough}", as verified by running the unit tests.

kkonstantine commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/Cast.java:
```diff
@@ -156,6 +159,8 @@ private R applyWithSchema(R record) {
             final Object origFieldValue = value.get(field);
             final Schema.Type targetType = casts.get(field.name());
             final Object newFieldValue = targetType != null ? castValueToType(field.schema(), origFieldValue, targetType) : origFieldValue;
+            log.trace("Cast field '{}' from '{}' to '{}'",
```
nit: Seeing the line right above, one would think this fits in one line

kkonstantine commented on core/src/main/scala/kafka/log/AbstractIndex.scala:
```diff
@@ -223,7 +223,13 @@ abstract class AbstractIndex[K, V](@volatile var file: File, val baseOffset: Lon
    */
   def deleteIfExists(): Boolean = {
-    closeHandler()
+    inLock(lock) {
```
Changes in this file seem completely irrelevant to this PR, unless an explanation is provided. Before diving more into this, is it intended to bring in these changes here? and if so, why? This also grants a comment in the commit log if these changes here are legit.

cyrusv replied:
Thanks for the catch, this was a sloppy rebase!

kkonstantine reviewed (commented):
I also left a few comments. Thanks @cyrusv

rhauch commented:
@kkonstantine, one more pass?

kkonstantine commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/TransformationChain.java:
```diff
@@ -40,6 +44,8 @@ public R apply(R record) {
         for (final Transformation<R> transformation : transformations) {
             final R current = record;
 
+            log.trace("Applying Single Message Transformation (SMT) {} to {}",
```
unnecessarily verbose for a log message IMO (see reply in previous thread)

kkonstantine commented:
@rhauch I replied and I'll wait for a couple clarifications still. Thanks. @cyrusv I've commented in previous threads.

cyrusv commented:
@kkonstantine thanks for the comments! They are addressed except I don't understand what you mean re:https://github.com/apache/kafka/pull/5860#discussion_r233664091.

cyrusv commented:
@kkonstantine thanks for the comments! I believe they are all addressed

kkonstantine reviewed (approved):
@cyrusv indeed. Nice improvement! 
LGTM by me.

cyrusv commented:
@ewencp Any more steps I should take before this is ready to merge?

ewencp commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/TransformationChain.java:
```diff
@@ -37,9 +41,11 @@ public TransformationChain(List<Transformation<R>> transformations, RetryWithTol
             final R current = record;
 
+            log.trace("Applying transformation {} to {}",
+                transformation.getClass().getSimpleName(), record);
```
I don't think you're going to want to use `getSimpleName()`. I think for a nested class that only returns the nested class name, and we use nested classes as a pattern with transformations to handle keys and values without duplicating a bunch of code, e.g. https://github.com/apache/kafka/blob/trunk/connect/transforms/src/main/java/org/apache/kafka/connect/transforms/ReplaceField.java#L194

cyrusv replied:
Good point -- I changed it back to the original `getName()`

ewencp commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/TransformationChain.java:
```diff
@@ -37,9 +41,11 @@ public TransformationChain(List<Transformation<R>> transformations, RetryWithTol
         if (transformations.isEmpty()) return record;
 
-        for (final Transformation<R> transformation : transformations) {
+        for (Transformation<R> transformation : transformations) {
```
why remove the `final` here?

kkonstantine replied:
@ewencp I suggested not using `final` in every new loop for consistency (several loops even here don't use it such as the one in `close`), but I didn't imply that we should change unaffected lines. 

In general in Connect my understanding is that we are not strict in demanding use of `final` in local variables. Let me know if something changed.

cyrusv replied:
Originally I had added `final` to the new code for-loop of the `toString()` method of the same file, and then removed it. I agreed with @kkonstantine's suggestion, and removed `final` from all the small for-loops of this class (to match the one already in `close()`. It looks like I stumbled into an inconsistency, already in the code, so I took out the change here as you suggest in order to keep this diff small and not conflate cleaning up consistency of "final" keyword with this logging change

ewencp replied:
@kkonstantine I don't feel strongly about it, but others do. Probably largely due to (with hindsight) the opposite default of what would have been ideal, i.e. would have been better to have to opt out of by-default immutable references rather than opt in.

We definitely already have some divergence within the broader AK codebase (`streams/` uses `final` more aggressively, not sure if it's checkstyle-enforced, Connect tends to use `final` mostly just for class fields). `clients` is a bit of a mix, but less `final` in `for` loops than not. And `core` is scala, so different story altogether.

I have no hard and fast rule here, main reason I even asked is because in this diff it is being removed from what was already there. No need to get hung up on it, but the other argument is to not change it in favor of minimizing diffs and maximizing mergability/cherry-pickability :) There are tradeoffs no matter which direction you choose....

kkonstantine replied:
Thanks @ewencp! Glad I'm still up-to-date on that. Happy to adjust per project and yes, unnecessary diffs are better to be skipped in non-cleanup/non-refactoring PRs. Huge fan of that.

ewencp commented on connect/api/src/main/java/org/apache/kafka/connect/connector/ConnectRecord.java:
```diff
@@ -140,7 +140,9 @@ public String toString() {
                 ", key=" + key +
+                ", keySchema=" + (keySchema != null ? "..." : "null") +
                 ", value=" + value +
+                ", valueSchema=" + (valueSchema != null ? "..." : "null") +
```
I would consider something other than just `...`. It took me some time to figure out why we were even adding this. `non-null` or something like that, or changing the format so it isn't `keySchema=` which implies (to me at least) that we're putting the actual value. Alternatively, you can include the actual schema, currently it's `toString()` just returns name & type, not a full, nested representation of the schema.

cyrusv replied:
I agree with your suggestion to use the `Schema.toString()`

cyrusv commented:
Thanks for the feedback @ewencp -- the issues are addressed!

ewencp commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/TransformationChain.java:
```diff
@@ -68,4 +74,11 @@ public int hashCode() {
     }
 
+    public String toString() {
+        StringJoiner chain = new StringJoiner(", ", "<" + getClass().getSimpleName() + ": ", ">");
```
i think the same comment about `getSimpleName` applies here (and in the body of this method) as well

ewencp reviewed (commented):
on last nit that i think will help logging clarity, after that ready to go

was this specifically for trunk or were there older branches where this additional logging would help debug issues? i.e., were you looking for back porting or just improving forward?

cyrusv commented:
Re:Version-targetting: There is a small rebase conflict against 2.0 (larger ones earlier than that), so I think 2.0 is a good target.  @ewencp

kkonstantine commented:
@cyrusv in AK the target is always `trunk`. I believe @ewencp asks whether he should backport after merging. It's indeed committer's responsibility to mention the actual target branch, but on github target is always `trunk`.

ewencp commented:
Oh yeah, sorry I wasn't clear. Generally the approach for target branches for AK is to target it at trunk unless it is a fix for a specific set of older branches, in which case you target to the relevant branch. If it's a change that should land on multiple branches, you just mention somewhere in the PR where you think it should go (and why if it wouldn't be clear, in this case it sounds like "as far back as is easy to port" since it's just to help with debugging).

The [merge script](https://github.com/apache/kafka/blob/trunk/kafka-merge-pr.py) helps with cherry-picking, and the committer can resolve any small issues along the way. (Or the committer can do cherry-picking themselves.) If the conflicts grow too large, they would just ask you to file a separate PR dealing with the conflicts and targeted at the older branch.

ewencp commented:
LGTM, merging to trunk, 2.1, and 2.0. If we want anything further back than that we can follow up with separate PRs.
